### PR TITLE
Implement torch.einsum (fixes #1889)

### LINKED ISSUE #1889: Support for einsum notation
As mentioned [here](https://discuss.pytorch.org/t/einstein-summation-in-pytorch/2492), support for [einsum notation](https://docs.scipy.org/doc/numpy/reference/generated/numpy.einsum.html) would be useful. It's in fact a feature that I am missing most after transitioning from TensorFlow. What's needed for the implementation is parsing the einsum string and deciding which PyTorch modules to use for which cases as well as automatically reshaping input tensor back and forth like done [here](https://github.com/tensorflow/tensorflow/blob/r1.2/tensorflow/python/ops/special_math_ops.py#L85).

### PR DESCRIPTION
**Pull request by t-vi**
This provides a simple implementation of einsum. It is built on top of the work for computing bilinear (#6110) and would obsolete PR #2008.
Properties:
- Implemented from scratch, but I used the tests from (@rockt)[https://gist.github.com/rockt/15ee013889d65342088e9260a377dc8f], including the fixme's.
- The obvious unsupported features are taking diagonals (einsum('ii->i',(a,)) and ellipses.

Implementation:
- It uses a naive left-to-right pairwise contractions at the moment.
- Backward works automagically.
- I had to complete the strings implementation in the python-ATen interface.

### CODE REVIEW

File: tools/autograd/gen_python_functions.py [hunk 1]
```diff
@@ -212,6 +212,7 @@ def create_python_bindings(python_functions, has_self, is_module=False):
         'int64_t': 'toInt64',
         'bool': 'toBool',
         'double': 'toDouble',
+        'std::string' : 'string',
```
Comment (fmassa) [inline tools/autograd/gen_python_functions.py]: nit: I think that once https://github.com/pytorch/pytorch/pull/6283 is merged, you will be able to use [`String`](https://github.com/pytorch/pytorch/pull/6283/files#diff-54816789e967b71e1be8048e53f8ccbdR29)

**Comment by fmassa** [comment]
Random thought: I have the impression that TensorComprehensions would be a good tool to generate efficient code for einsum

**Comment by t-vi** [comment]
I'll clean up the mess about bilinear and flake8 later. Sorry.

File: tools/autograd/gen_python_functions.py (hunk 1 above)
Comment (t-vi) [reply tools/autograd/gen_python_functions.py]: Thak you for the pointer! I will rebase on that eventually.

**Comment by rockt** [comment]
This looks great! I'd love to see it merged soon.

**Comment by t-vi** [comment]
So with #6479 (which has code even if it is not a PR yet), we can also handle traces. What would make the reviewer's job easiest, to amend the code here now or to wait for this to be merged and send a second PR?
If the latter, I think this is basically ready for review.

**Review by fmassa — COMMENTED** [review]
I did a first pass and have a few minor comments, but this looks good overall.
I think this needs a rebase though.

Also, can you add a few more comments in einsum, explaining the order of which you process the equation and how it corresponds to the permutations / `sumproduct_pair`?
Thanks!

File: aten/src/ATen/ScalarType.h [hunk 2]
```diff
@@ -143,5 +143,6 @@ static inline ScalarType promoteTypes(ScalarType a, ScalarType b) {
 struct Tensor;
 typedef ArrayRef<int64_t> IntList;
 typedef ArrayRef<Tensor> TensorList;
+typedef std::string String;
```
Comment (apaszke) [inline aten/src/ATen/ScalarType.h]: Can you please remove this typedef? You should be able to use `std::string` in every place where you used `String`

File: aten/src/ATen/native/Linear.cpp [hunk 3]
```diff
@@ -1,9 +1,202 @@
...
+  return result;
+}
+
+
+Tensor einsum(String eqn, TensorList tensors) {
+  std::string in_eqn;
+  size_t pos;
+  std::vector<std::int64_t> number_of_occurences(26, 0);
+  std::vector<std::int64_t> last_occurence(26, -1);
+  std::vector<std::int64_t> sorted_position(26, -1);
```
Comment (apaszke) [inline aten/src/ATen/native/Linear.cpp]: What's 26? Also, can't we use `std::array` if the size is constant? That enables many more compiler optimizations.

Also, do we need to initialize them? `sorted_position` and `last_occurence` sound like lookup tables that are mostly unused

File: aten/src/ATen/native/Linear.cpp [hunk 4]
```diff
@@ -1,9 +1,202 @@
...
+  std::stringstream eqn_stream(in_eqn);
+  std::string term;
+  while (! eqn_stream.eof()) {
+    std::getline(eqn_stream, term, ',');
+    int64_t dims_in_operand = 0;
+    for (auto &c : term) {
+      AT_ASSERT(('a' <= c) && (c <= 'z'), "only lowercase letters a-z allowed as indices");
+      int64_t index_num = c-'a';
+      number_of_occurences[index_num]++;
+      AT_ASSERT(last_occurence[index_num] < operand, "diagonals (multiple occurences of the same index for one tensor) not implemented yet")
```
Comment (apaszke) [inline aten/src/ATen/native/Linear.cpp]: This doesn't seem like a robust check. What if you do: `ab,cd,bb`? It only checks that the average number of appearances in all prefixes in `<=1`

File: aten/src/ATen/native/Linear.cpp [hunk 5]
```diff
@@ -1,9 +1,202 @@
...
+  if (! keepdim) {
+    for (int i = dim-1; i>=0; i--)
+      if (sum_dims[i])
+	result.squeeze_(i);
+  }
+  return result;
+}
+
+
+Tensor einsum(String eqn, TensorList tensors) {
```
Comment (apaszke) [inline aten/src/ATen/native/Linear.cpp]: Would it be possible to split this one into separate functions that do the parsing, preprocessing and actual compute?

File: aten/src/ATen/native/Linear.cpp [hunk 6]
```diff
@@ -1,9 +1,202 @@
...
+  std::string in_eqn;
+  size_t pos;
+  std::vector<std::int64_t> number_of_occurences(26, 0);
+  std::vector<std::int64_t> last_occurence(26, -1);
+  std::vector<std::int64_t> sorted_position(26, -1);
+  if ((pos = eqn.find("->")) != std::string::npos) {
```
Comment (apaszke) [inline aten/src/ATen/native/Linear.cpp]: What does an expression without a `->` mean?

File: aten/src/ATen/native/Linear.cpp (hunk 3 above)
Comment (apaszke) [reply aten/src/ATen/native/Linear.cpp]: Also, comments on what will be contained in those arrays? Possibly with an example input, and their contents in this particular case

File: aten/src/ATen/native/Linear.cpp [hunk 7]
```diff
@@ -1,9 +1,202 @@
...
+      dims_in_operand++;
+    }
+    AT_ASSERT((int64_t) tensors.size()>operand, "more operands in equation than tensors");
+    AT_ASSERT(dims_in_operand == tensors[operand].dim(), "dimension mismatch for operand %zd: equation %zd, tensor %zd", operand, dims_in_operand, tensors[operand].dim());
+    operand++;
+  }
+  AT_ASSERT((int64_t) tensors.size()==operand, "more tensors than operands in equation");
+
+  int64_t num_outputs = 0;
+  std::vector<int64_t> position_labels;
```
Comment (apaszke) [inline aten/src/ATen/native/Linear.cpp]: Same here. A comment would be useful

File: aten/src/ATen/native/Linear.cpp [hunk 8]
```diff
@@ -1,9 +1,202 @@
...
+  if (pos != std::string::npos) {
+    for (auto &c : eqn.substr(pos+2)) {
+      AT_ASSERT(('a' <= c) && (c <= 'z'), "only lowercase letters a-z allowed as indices");
+      int64_t index_num = c-'a';
+      AT_ASSERT(sorted_position[index_num] == -1, "index %c occurs twice in output", c);
+      sorted_position[index_num] = num_outputs;
+      position_labels.push_back(index_num);
+      num_outputs++;
```
Comment (apaszke) [inline aten/src/ATen/native/Linear.cpp]: There's only a single output, can you please rename this to `num_rhs_iterators`?

File: aten/src/ATen/native/Linear.cpp [hunk 9]
```diff
@@ -1,9 +1,202 @@
...
+      sorted_position[idx] = position;
+      position_labels.push_back(idx);
+      position++;
+    }
+  }
+  std::vector<Tensor> permuted_ops;
+  eqn_stream.clear();
+  eqn_stream.seekg(0, std::ios_base::beg);
+  for (int64_t op = 0; op < (int64_t) tensors.size(); op++) {
+    std::vector<int64_t> axes(26, -1);
```
Comment (apaszke) [inline aten/src/ATen/native/Linear.cpp]: Same here. Can it be an array, and does it really need to be initialized?

File: aten/src/ATen/native/Linear.cpp [hunk 10]
```diff
@@ -1,9 +1,202 @@
...
+    if ((number_of_occurences[idx] > 0) && (sorted_position[idx]==-1)) {
+      sorted_position[idx] = position;
+      position_labels.push_back(idx);
+      position++;
+    }
+  }
+  std::vector<Tensor> permuted_ops;
+  eqn_stream.clear();
+  eqn_stream.seekg(0, std::ios_base::beg);
+  for (int64_t op = 0; op < (int64_t) tensors.size(); op++) {
```
Comment (apaszke) [inline aten/src/ATen/native/Linear.cpp]: A comment saying what this part does? It seems to reorder the dimensions somehow according to the indices, can you please explain how?

File: aten/src/ATen/native/Linear.cpp [hunk 11]
```diff
@@ -1,9 +1,202 @@
...
+	permuted_ops.back().unsqueeze_(dim);
+      }
+    }
+  }
+  // we just go left to right. numpy allows to optimize the path...
+  Tensor result = permuted_ops[0];
+  for (int64_t idx = 0; idx < 26; idx++) {
+    if ((last_occurence[idx] == 0)
+	&& (sorted_position[idx]>=num_outputs)) {
+      result = result.sum(sorted_position[idx], true);
```
[the remark this answers is not in the document]
Comment (apaszke) [reply aten/src/ATen/native/Linear.cpp]: Yep, let's wait until the multi-dim sum gets in, and use it there!

File: aten/src/ATen/native/Linear.cpp (hunk 3 above)
Comment (fmassa) [reply aten/src/ATen/native/Linear.cpp]: That was my initial thoughts as well about 26, which is the number of letters in the alphabet (between a and z)

File: aten/src/ATen/native/Linear.cpp (hunk 8 above)
Comment (t-vi) [reply aten/src/ATen/native/Linear.cpp]: `num_output_dims` or `output_dims` or so? It really is the number of dimensions in the output.

File: aten/src/ATen/native/Linear.cpp [hunk 12]
```diff
@@ -1,9 +1,202 @@
...
+    for (auto &c : position_labels) {
+      if (axes[c] > -1) {
+	permutation.push_back(axes[c]);
+      }
+    }
+    permuted_ops.push_back(tensors[op].permute(permutation));
+    for (int64_t dim = 0; dim < (int64_t) position_labels.size(); dim++) {
+      auto c = position_labels[dim];
+      if (axes[c] == -1) {
+	permuted_ops.back().unsqueeze_(dim);
```
[the remark this answers is not in the document]
Comment (t-vi) [reply aten/src/ATen/native/Linear.cpp]: The indices are dimensions in the result after unsqueezing, so left to right is on purpose.

File: test/test_torch.py [hunk 13]
```diff
@@ -1296,6 +1296,57 @@ def test_cmul(self):
...
+            ("ij,kj->ik", A, B),        # matmul
+            ("ij,ab->ijab", A, E),      # matrix outer product
+            # -- Tensor
+            ("aij,ajk->aik", C, D),     # batch matmul
+            ("ijk,jk->i", C, A),        # tensor matrix contraction
+            ("aij,jk->aik", D, E),      # tensor matrix contraction
+            ("abcd,dfg->abcfg", F, G),  # tensor tensor contraction
+            ("ijk,jk->ik", C, A),       # tensor matrix contraction with double indices
+            ("ijk,jk->ij", C, A),       # tensor matrix contraction with double indices
+            ("ijk,ik->j", C, B),        # fixme -- non contiguous
```
[the remark this answers is not in the document]
Comment (t-vi) [reply test/test_torch.py]: Ha, of course they are fixed. :)

File: torch/_torch_docs.py [hunk 14]
```diff
@@ -1489,6 +1489,67 @@
...
+    operands (list of Tensors): The operands to compute the Einstein sum of.
+           Note that the operands are passed as a list, not as individual arguments.
+
+Examples::
+
+    >>> x = torch.randn(5)
+    >>> y = torch.randn(4)
+    >>> torch.einsum('i,j->ij', (x,y))  # outer product
+
+    -1.0066 -2.0433 -0.8290  0.8429
```
[the remark this answers is not in the document]
Comment (t-vi) [reply torch/_torch_docs.py]: Maybe we should regenerate *all* documentation examples regularly (and set a fixed seed before running them)... Then we'd be sure they are up to date and working. I could send a PR for that it is helpful - the documentation is very dear to me.

File: aten/src/ATen/WrapDimUtilsMulti.h [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+#include <bitset>
+
+namespace at {
+
+// This is in an extra file to work around strange interaction of
+// bitset on Windows with operator overloading
+
+constexpr size_t dim_bitset_size = 64;
+
+static inline std::bitset<dim_bitset_size> dim_list_to_vector(IntList dims, int64_t ndims, bool wrap_scalar=true) {
```
[the remark this answers is not in the document]
Comment (t-vi) [reply aten/src/ATen/WrapDimUtilsMulti.h]: No, I just need to pass it to the 1d-case.

File: aten/src/ATen/native/Linear.cpp (hunk 4 above)
Comment (t-vi) [reply aten/src/ATen/native/Linear.cpp]: note that this is last_occurence in the assert, not number_of_occurences (hmhm. it should be occurrence with two r).

File: aten/src/ATen/native/Linear.cpp (hunk 3 above)
Comment (t-vi) [reply aten/src/ATen/native/Linear.cpp]: It is the letters in the alphabet, I changed 26 to a constexpr to be reduce confusing magic.

File: aten/src/ATen/native/Linear.cpp (hunk 5 above)
Comment (t-vi) [reply aten/src/ATen/native/Linear.cpp]: I don't have strong feelings about it. It would need to pass around a lot of information.
Also, I go through the string twice. I can change that by caching indices in a vector-vector-int64, but I'm not 100% certain that that is a lot better.

File: aten/src/ATen/native/Linear.cpp (hunk 6 above)
Comment (t-vi) [reply aten/src/ATen/native/Linear.cpp]: I added a comment for that, but it means "indices occurring only once, alphabetically sorted". It's a tad exotic, but I'd include it for compatibility.

File: torch/_torch_docs.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (fmassa) [reply torch/_torch_docs.py]: I agree, and it's being discussed in https://github.com/pytorch/pytorch/issues/6662

**Comment by fmassa** [comment]
I think there are also some submodules that were updated, is it intentional?

**Comment by t-vi** [comment]
No, and I'm scratching my head how to get rid of it. :(

Am 17. April 2018 20:45:53 MESZ schrieb Francisco Massa <notifications@github.com>:

File: torch/_torch_docs.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (zafarali) [reply torch/_torch_docs.py]: Is there a reason we need to pass arguments as tuples?

Btw great work!

[the remark this answers is not in the document]
Comment (t-vi) [reply torch/_torch_docs.py]: Is there a reason we need to pass arguments as tuples?

Great comment, thank you!
The parsing of arguments is in [`torch/csrc/utils/python_arg_parser.cpp`, in `FunctionSignature::parse`](https://github.com/pytorch/pytorch/blob/187955b9599c162f75d19ceb65fde02bd85253b3/torch/csrc/utils/python_arg_parser.cpp#L363). Varargs are only supported if there is a single parameter and that is of type IntList. This is because in the arg parsing [the entire argument tuple is assigned as the argument](https://github.com/pytorch/pytorch/blob/187955b9599c162f75d19ceb65fde02bd85253b3/torch/csrc/utils/python_arg_parser.cpp#L409). Now, if we wanted to have a slice there (which would be the way to support it generically), we would need to construct one using `PyTuple_GetSlice`. But that would be a new reference, and suddenly we would need to release the reference in the indirect caller - who typically used `PythonArgParse.parse` . So the way to support it would be to return something like "vararg-tuple" and deallocate that or have `PythonArgs` do that automatically when it goes out of scope - but I'm not sure whether the latter is a good idea if the caller does have the `ParsedArgs` struct as well. All this seems to be fairly invasive and also I don't understand the rationale of exposing `ParsedArgs` to callers (rather than only showing the `PythonArgs`) to not have that be a fairly strong distraction to the einsum implementation.

The obvious alternative is to do our own arg parsing and blacklist it in automatic generation - similar to what is [done for clamp](https://github.com/pytorch/pytorch/blob/187955b9599c162f75d19ceb65fde02bd85253b3/tools/autograd/templates/python_torch_functions.cpp#L51). The additional complication would be that we cannot use `PythonArgParse` - not that bad if we can still use `PythonArgs`.

As such, I would tentatively not want to burden the reviewers with this. However if they would like to review that along with the other bits, I would be happy to include that in the PR as well.

**Comment by fmassa** [comment]
There is still a conflict with `third_party/onnx`, but after that I think this is good to merge!

**Comment by fmassa** [comment]
And I think it's fine to have the tensor arguments being passed as tuple for the moment.
Another workaround for the tuple parsing would be to have the python do the tuple wrapping:
```python
def einsum(eqn, *args):
    return torch._einsum(eqn, args)
```
but I don't think this is really necessary. @apaszke any thoughts on that? Do we **really** want to follow numpy einsum interface?

**Comment by apaszke** [comment]
@fmassa can you elaborate? What's the issue with the NumPy interface?

**Review by apaszke — APPROVED** [review]
The comments are super helpful, and the code looks great now. Thanks a lot @t-vi!

**Comment by apaszke** [comment]
@fmassa Not sure. It would be nice to allow this, but I don't think the tuple is too bad. Let's just get this in as is.

**Comment by fmassa** [comment]
@apaszke there is no issue with the tuple interface, is just that according to Thomas comment in https://github.com/pytorch/pytorch/pull/6307#issuecomment-382322053 , we would need to either modify the parser to handle the TensorList in the end, or go through the extra python dispatch layer. But maybe I'm missing something?

File: torch/_torch_docs.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (fmassa) [reply torch/_torch_docs.py]: Let's get this as in!

**Review by fmassa — APPROVED** [review]
